modules/host/announce.go

```diff
@@ -63,26 +68,22 @@ func (h *Host) announce(addr modules.NetAddress) error {
 // arbitrary data, signing the transaction, and submitting it to the
 // transaction pool.
 func (h *Host) Announce() error {
+	h.mu.Lock()
+	defer h.mu.Unlock()
 	h.resourceLock.RLock()
 	defer h.resourceLock.RUnlock()
 	if h.closed {
 		return errHostClosed
 	}
 
-	// Get the external IP again; it may have changed.
-	h.learnHostname()
-	h.mu.RLock()
-	addr := h.netAddress
-	h.mu.RUnlock()
-
-	// Check that the host's ip address is known.
-	if addr.IsLoopback() && build.Release != "testing" {
-		return errors.New("can't announce without knowing external IP")
+	// Determine whether to use the settings.NetAddress or autoAddress.
+	if h.settings.NetAddress != "" {
+		return h.announce(h.settings.NetAddress)
 	}
-
-	h.mu.Lock()
-	defer h.mu.Unlock()
-	return h.announce(addr)
+	if (h.autoAddress == "" || h.autoAddress.IsLoopback()) && build.Release != "testing" {
```

Comment:
I was wondering this too. Supposedly the hostname discovery will not return a loopback address, so I think I can remove that clause.